packages/yoroi-extension/chrome/extension/utils.js

```diff
@@ -0,0 +1,135 @@
+  openedWindow: boolean,
+  // resolve function from signRequest's Promise
+  resolve: any
+|}
+
+// This is a temporary workaround to DB duplicate key constraint violations
+// that is happening when multiple DBs are loaded at the same time, or possibly
+// this one being loaded while Yoroi's main App is doing DB operations.
+let loadedDB: ?lf$Database = null;
+let dbPromise: ?Promise<lf$Database> = null;
+
+export async function loadDB(): Promise<lf$Database> {
+  if (loadedDB == null) {
+    if (dbPromise == null) {
+      dbPromise = loadLovefieldDB(schema.DataStoreType.INDEXED_DB)
+        .then(db => {
+          loadedDB = db;
+          return Promise.resolve(loadedDB);
+        });
+    }
+    return dbPromise;
+  }
+  return Promise.resolve(loadedDB);
+}
+
+export type AccountIndex = number;
+
+// AccountIndex = successfully connected - which account the user selected
+// null = refused by user
+type ConnectedStatus = ?AccountIndex | {|
+  // response (?AccountIndex) - null means the user refused, otherwise the account they selected
+  resolve: any,
+  // if a window has fetched this to show to the user yet
+  openedWindow: boolean,
+|};
+
+export type ConnectedSite = {|
+  url: string,
+  status: ConnectedStatus,
+  pendingSigns: Map<RpcUid, PendingSign>
+|};
+
+async function getChecksum(
+  publicDeriver: ReturnType<typeof asGetPublicKey>,
+): Promise<void | WalletChecksum> {
+  if (publicDeriver == null) return undefined;
+
+  const hash = (await publicDeriver.getPublicKey()).Hash;
+
+  const isLegacyWallet =
+    isCardanoHaskell(publicDeriver.getParent().getNetworkInfo()) &&
+    publicDeriver.getParent() instanceof Bip44Wallet;
+  const checksum = isLegacyWallet
+    ? legacyWalletChecksum(hash)
+    : walletChecksum(hash);
+
+  return checksum;
+}
+
+export async function getWalletsInfo(): Promise<AccountInfo[]> {
+  try {
+    const db = await loadDB();
+    const wallets = await getWallets({ db });
+    // information about each wallet to show to the user
+    const accounts = [];
+    for (const wallet of wallets) {
+      const conceptualWallet = wallet.getParent();
+      const withPubKey = asGetPublicKey(wallet);
+
+      const conceptualInfo = await conceptualWallet.getFullConceptualWalletInfo();
+      if (isErgo(conceptualWallet.getNetworkInfo())) {
+        // TODO: we can't get the pending txs from background.js from here
+        // since it runs in a different context. Need a better solution
+        const pendingTxs = [];
+        const balance = await connectorGetBalance(wallet, pendingTxs, 'ERG');
+        accounts.push({
+          name: conceptualInfo.Name,
+          balance: balance.toString(),
+          checksum: await getChecksum(withPubKey)
+        });
+      }
+    }
+    return accounts;
+  } catch (error) {
+    throw new GenericApiError();
+  }
+}
+
+export function getActiveSites(): Array<string> {
+  const activeSites = []
+  // TODO: can't get connected sites from background.js
```

Comment:
Will need to think of a better way to get this information since the background page is the one managing this. Maybe this will have to be passed as a message or passed through storage.

Reply:
When we go to manifest v3 we'll probably end up storing it in local storage since you can't have global variables in the new web workers, but if they change the port API we might not need to.